File: Tree/boundary_traversal.rb
Node = Struct.new(:key, :left, :right)

def print_boundary(root)
  return if root == nil

  print "#{root.key} "
  print_left_boundary(root.left)
  print_leafs(root)
  print_right_boundary(root.right)
end

def print_left_boundary(node)
  return if node == nil || (node.left == nil && node.right == nil)
  print "#{node.key} "
  print_left_boundary(node.left || node.right)
end

def print_leafs(node)
  return if node == nil

  if node.left == nil and node.right == nil
    print "#{node.key} "
    return
  end
  print_leafs(node.left)
  print_leafs(node.right)
end

def print_right_boundary(node)
  return if node == nil || (node.left.nil? && node.right.nil?)
  print_right_boundary(node.right || node.left)
  print "#{node.key} "
end

root =  Node.new(20)
root.left =  Node.new(8)
root.left.left =  Node.new(4)
root.left.right =  Node.new(12)
root.left.right.left =  Node.new(10)
root.left.right.right =  Node.new(14)
# root.right =  Node.new(22)
# root.right.right =  Node.new(25)
print_boundary(root)
puts

File: Tree/BoundaryTraversal.java
import java.util.*;

/*
	https://practice.geeksforgeeks.org/problems/boundary-traversal-of-binary-tree/1#
*/
class BoundaryTraversal {
	public static void main(String[] args) {
		Node root = new Node(20);
		root.left = new Node(8);
		root.right = new Node(22);
		root.right.right = new Node(25);

		root.left.left = new Node(4);
		root.left.right = new Node(12);
		root.left.right.left = new Node(10);
		root.left.right.right = new Node(14);

		System.out.println(new Solution().printBoundary(root)); // 20 8 4 10 14 25 22
	}
}


class Solution {

	ArrayList <Integer> printBoundary(Node node) {

	    ArrayList<Integer> res = new ArrayList<>();

        res.add(node.data);

        if (node.left == null && node.right == null)
            return res;

	    if(node.left != null)
	       leftBoundary(node.left, res);

	    allLeafs(node, res);

	    if (node.right != null)
	        rightBoundary(node.right, res);

	    return res;
	}

	private void rightBoundary(Node node, ArrayList<Integer> res) {
	    if (node == null || (node.left == null && node.right == null)) return;

	    if (node.right != null)
	        rightBoundary(node.right, res);
        else
            rightBoundary(node.left, res);

        res.add(node.data);
	}

	private void allLeafs(Node node, ArrayList<Integer> res) {
	    if (node == null) {
	        //
	    } else if (node.left == null && node.right == null) {
	        res.add(node.data);
	    } else {
	        allLeafs(node.left, res);
	        allLeafs(node.right, res);
	    }
	}

	private void leftBoundary(Node node, ArrayList<Integer> res) {
	    if (node == null || (node.left == null && node.right == null)) return;

	    res.add(node.data);

	    if(node.left != null) leftBoundary(node.left, res);
	    else leftBoundary(node.right, res);

	}
}

class Node  {
    int data;
    Node left, right;
    Node(int d)  { data = d; }
}
